text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,184 @@
+Use `bundle lock` to add checksums to Gemfile.lock.
+```
+
+When a gem version is updated in the lockfile using a source that has checksums (like rubygems.org), the checksum should be recorded. Running `bundle update` or `bundle add` should record the checksum from the source into the Gemfile.lock. If a checksum is not available from the source because the source does not provide such info, then a checksum should be created during install using the .gem file or a warning should be printed.
+
+If a user wishes to ensure that the environment they are using contains only gems installed from verified sources that match the checksums in the Gemfile.lock, the command `bundle pristine` should be run. The `bundle pristine` command installs every gem fresh from the .gem source, but does not fetch gems from the server or read the source index. Therefore, if a users wants all gems to be verified upon install, `bundle pristine` should be run after checksums are written to the Gemfile.lock.
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems are missing checksums. Source integrity could not be verified.
+Use `bundle lock` to add checksums to Gemfile.lock.
+$ bundle lock --update
+Writing lockfile to path/to/Gemfile.lock
+$ bundle pristine
+Installing rake 13.0.6 (verified)
+…
+Installing rspec 3.12.0 (verified)
+$
+```
+
+If a gem that is about to be installed does not match the checksum, an error is generated.
+
+Example:
+
+```
+$ bundle install
+Installing rake 13.0.6
+Bundler cannot continue installing rake-13.0.6.
+The checksum for the downloaded `rake-13.0.6.gem` does not match the known checksum for the gem. 
+
+Expected: sha256-814828c34f1315d7e7b7e8295184577cc4e969bad6156ac069d02d63f58d82e8 from: Gemfile.lock:21:1 CHECKSUMS rake (13.0.6)
+Downloaded: sha256-69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b6 from:
+path/to/rake-13.0.6.gem SHA256 hexdigest
+
+To resolve this issue:
+Delete the downloaded gem files referenced above and run `bundle install`
+
+If you are sure that the downloaded gem is correct, you can run…
+(RFC TODO: Currently undecided on what the command should be here to allow unverified gems.)
+`bundle install --skip-verification` OR
+`bundle config set unverified ignore` OR
+Remove the offending line from CHECKSUMS in Gemfile.lock
+```
+
+Certain checksums will always be unavailable because the source does not provide a checksum. When a checksum is not available, a line can be added to the Gemfile.lock CHECKSUMS section that explicitly indicates that no checksum is recorded for that gem version. Future installs of gems with intentionally ignored checksums should not error or warn when the gem is installed. The **(RFC TODO)** command will allow idicating that a gem version should ignore checksum verifications. Alternatively, the SHA256 digest of the gem at install time can be recorded allowing the user to verify future installs do not deviate from the recorded checksum.
```

Comment:
> Alternatively, the SHA256 digest of the gem at install time can be recorded allowing the user to verify future installs do not deviate from the recorded checksum.

I like this, it matches with the `lock on first install` approach bundler already takes with dependency resolution

Reply:
Great. This is how it is implemented right now for .gem files.

There's still git or path gems that will never have checksums. I don't think we need any explicit ignore for them, as I started with at the top. I can update that.